keyedcache/api.go

```diff
@@ -0,0 +1,153 @@
+// is created by NewObject() this interface is initialized).
+//
+// CachedObject supports the following methods. These may be invoked by a client
+// that has a hold on the object.  Successful lookup of an object in the cache
+// returns a pointer to a held object.
+//
+//   o Hold() -- get an additional hold on the object
+//   o Release() -- release a hold on the object; once all holds on an object
+//     are released the object may be invalidated
+//   o MarkDirty() -- mark a held object dirty; dirty objects are cleaned
+//     (written) by the cache before invalidation unless a forced invalidation
+//     is done
+//   o IsDirty() -- returns true if the object is dirty
+//   o Flush() -- cause a dirty object to be flushed via a call to Write();
+//     panics if the object is not dirty
+//
+// If Flush() fails because the corresponding Write() fails then the object is
+// marked dirty.
+//
+type CachedObject interface {
+	Hold(cntxt *ReqContext)
+	Release(cntxt *ReqContext)
+	MarkDirty(cntxt *ReqContext)
+	IsDirty(cntxt *ReqContext) bool
+	Flush(cntxt *ReqContext) (err error)
+}
+
+// Cache caches objects identified by keys, upto a specified maximum amount of
+// memory.  Objects in the cache are accessed by calling Lookup() with the
+// desired key and Lookup() returns a pointer to an object with a hold on it.
+// When the caller is done with the object it must release the hold.
+//
+// If the object is not currently cached, Cache will create a new object
+// corresponding to the key or reuse an existing one.  It is initialized by a
+// call to Object.Read() before return.  Cache locks the object duing the call
+// to Object.Read() so only one object with a given key is present in the cache
+// and no thread can access it until it is initialized.
+//
+// There is no provision for otherwise locking the object.  In particular, an
+// object can be marked dirty or flushed while multiple threads have a hold on
+// the object.
+//
+// A Cache provides the following methods:
+//
+//   o Lookup() -- lookup an object by key, creating it if necessary; may return
+//     an error if Read() fails; blocks if the cache is full
+//   o Flush() -- flush all objects matched by the function predicate(); return
+//     an error if any of the Object.Write() fails
+//   o Inval() -- invalidate all  objects matched by the function predicate();
+//     InvalFlush determines whether dirty objects are flushed before they are
+//     invalidated; Inval() will block until a held object is released; Inval()
+//     will return an error if flush failed
+//
+type Cache interface {
+	Lookup(cntxt *ReqContext, key Key) (err error)
+	Flush(cntxt *ReqContext, predicate func(cntxt *ReqContext, key *Key) bool) (err error)
+	Inval(cntxt *ReqContext, predicate func(cntxt *ReqContext, key *Key) bool,
+		flushIfDirty InvalFlush) (err error)
+	Free(cntxt *ReqContext)
+}
+
+type InvalFlush int
+
+const (
+	INVAL_DIRTY InvalFlush = iota
+	FLUSH_DIRTY InvalFlush = iota
```

Comment:
I'm going to assume you want INVAL_DIRTY and FLUSH_DIRTY to have the same (zero) value here?
If not, I think you wanted to remove the second "= iota"... thus emulating an "enum" sequence that Go otherwise lacks.